presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionTask.java

```diff
@@ -0,0 +1,267 @@
+            this.taskInfoCodec = taskInfoCodec;
+        }
+    }
+
+    public TaskInfo getTaskInfo()
+    {
+        return taskInfo;
+    }
+
+    public void start()
+    {
+        sendUpdateRequest();
+        scheduleGetTaskInfo();
+    }
+
+    public void stop()
+    {
+        if (scheduledFuture != null) {
+            scheduledFuture.cancel(false);
+        }
+    }
+
+    private void sendUpdateRequest()
+    {
+        Optional<byte[]> fragment = Optional.of(planFragment.toBytes(planFragmentCodec));
+        Optional<TableWriteInfo> writeInfo = Optional.of(tableWriteInfo);
+        TaskUpdateRequest updateRequest = new TaskUpdateRequest(
+                session.toSessionRepresentation(),
+                session.getIdentity().getExtraCredentials(),
+                fragment,
+                sources,
+                outputBuffers,
+                writeInfo);
+        byte[] taskUpdateRequestJson = taskUpdateRequestCodec.toBytes(updateRequest);
+
+        HttpUriBuilder uriBuilder = uriBuilderFrom(taskInfo.getTaskStatus().getSelf());
+        Request request = setContentTypeHeaders(false, preparePost())
+                .setUri(uriBuilder.build())
+                .setBodyGenerator(createStaticBodyGenerator(taskUpdateRequestJson))
+                .build();
+
+        ResponseHandler responseHandler = createAdaptingJsonResponseHandler(taskInfoCodec);
+        ListenableFuture<BaseResponse<TaskInfo>> future = httpClient.executeAsync(request, responseHandler);
+        FutureCallback callback = new SimpleHttpResponseHandler<>(new UpdateResponseHandler(), request.getUri(), stats.getHttpResponseStats(), NATIVE_EXECUTION_TASK_ERROR);
+
+        Futures.addCallback(
+                future,
+                callback,
+                executor);
+    }
+
+    private void scheduleGetTaskInfo()
+    {
+        scheduledFuture = updateScheduledExecutor.scheduleWithFixedDelay(() -> {
+            try {
+                HttpUriBuilder uriBuilder = uriBuilderFrom(taskInfo.getTaskStatus().getSelf());
+                Request request = setContentTypeHeaders(false, prepareGet())
+                        .setUri(uriBuilder.build())
+                        .build();
+
+                ResponseHandler responseHandler = createAdaptingJsonResponseHandler(taskInfoCodec);
+                ListenableFuture<BaseResponse<TaskInfo>> future = httpClient.executeAsync(request, responseHandler);
+                FutureCallback callback = new SimpleHttpResponseHandler<>(new TaskInfoHandler(), request.getUri(), stats.getHttpResponseStats(), NATIVE_EXECUTION_TASK_ERROR);
+
+                Futures.addCallback(
+                        future,
+                        callback,
+                        executor);
+            }
+            catch (Throwable t) {
+                throw t;
+            }
+        }, 0, GET_TASK_INFO_INTERVALS, MILLISECONDS);
+    }
+
+    private synchronized void updateTaskInfo(TaskInfo newValue)
```

Comment:
synchronizing this method does not help to protect taskInfo retrieval in NativeTaskExecution. Since you are accessing the same taskInfo object in multiple objects then can we do following whenever you try to use taskInfo:

`
synchronized (taskInfo) {
  taskInfo.blah();
  ...
}
`